open_instruct/grpo_fast.py

```diff
@@ -1152,7 +1155,11 @@ def generate_with_engines(prompts: List[List[int]], sampling_params: SamplingPar
         inference_results_Q.put((response_ids, finish_reasons))
 
         # Evaluate the model
-        if eval_prompt_token_ids is not None and (training_step - 1) % eval_freq == 0:
+        if (
+            eval_prompt_token_ids is not None
+            and num_evals > 0
+            and ((training_step - 1) % eval_freq == 0 or training_step == num_training_steps)
+        ):
```

Comment:
One thing that may also help is to set `evaluation_inference_results_Q = Queue(maxsize=10)` and also a timeout = 100 seconds in the final check (instead of a None that could block forever). Was gonna add just to be on the safe side, but what you have here should also work!

Reply:
Oh yeah, we should add some timeout anyway, good point.